corehq/util/decorators.py

```diff
@@ -62,3 +64,69 @@ def __enter__(self):
 
     def __exit__(self, exc_type, exc_val, exc_tb):
         pass
+
+
+class CouldNotAqcuireLock(Exception):
+    pass
+
+
+# Sorry this is so magic
+def _get_unique_key(format_str, fn, *args, **kwargs):
+    """
+    Lines args and kwargs up with those specified in the definition of fn and
+    passes the result to `format_str.format()`.
+    """
+    varnames = fn.func_code.co_varnames
+    kwargs.update(dict(zip(varnames, args)))
+    return ("{}-" + format_str).format(fn.__name__, **kwargs)
+
+
+def locking_task(unique_key, default_retry_delay=30, timeout=5*60, max_retries=3):
+    """
+    Define a locking celery task to prevent race conditions.
+
+    :param unique_key: string used to lock the task.  There will be one lock
+        per unique value.  You may use any arguments that will be passed to the
+        function.  See example.
+    :param default_retry_delay: seconds to wait before retrying if a lock is
+        encountered
+    :param timeout: timeout on the lock (in seconds).  Normally the lock should
+        be released when the task completes, but you should also define a
+        timeout in case something goes wrong.  This must be greater than the
+        maximum length of the task.
+
+    Usage:
+        @locking_task("{user.username}-{from}", default_retry_delay=2)
+        def greet(user, from="Dimagi"):
+            ...
+
+        greet.delay(joeshmoe)
+        # Locking key used would be "greet-joeshmoe@test.commcarehq.org-Dimagi"
+    """
+    def decorator(fn):
+        # register task with celery.  Note that this still happens on import
+        @task(bind=True, queue='background_queue', ignore_result=True,
+              default_retry_delay=default_retry_delay, max_retries=max_retries)
+        def _inner(self, *args, **kwargs):
+            if settings.UNIT_TESTING:  # Don't depend on redis
```

Comment:
just curious why this is necessary - even though we run tasks synchronously during tests it seems the behavior would still be the same if we left it in (acquire lock, run task, release lock). is this just to prevent the overhead of doing that? if so i think it would be better to check for CELERY_ALWAYS_EAGER (and should check for it using getattr(settings, ...) since it might not be defined in settings)